src/datatablerow/utils.jl

```diff
@@ -0,0 +1,190 @@
+    @inbounds for i in eachindex(h)
+        h[i] = hash(v[i], h[i])
+    end
+    h
+end
+
+function hashrows_col!{T}(h::Vector{UInt}, v::AbstractVector{T})
+    @inbounds for i in eachindex(h)
+        h[i] = isnull(v[i]) ?
+               hash(NULL_MAGIC, h[i]) :
+               hash(v[i], h[i])
+    end
+    h
+end
+
+# should give the same hash as AbstractVector{T}
+function hashrows_col!{T}(h::Vector{UInt}, v::AbstractCategoricalVector{T})
+    # TODO is it possible to optimize by hashing the pool values once?
+    @inbounds for (i, ref) in enumerate(v.refs)
+        h[i] = hash(CategoricalArrays.index(v.pool)[ref], h[i])
+    end
+    h
+end
+
+# should give the same hash as AbstractNullableVector{T}
+function hashrows_col!{T}(h::Vector{UInt}, v::AbstractNullableCategoricalVector{T})
+    # TODO is it possible to optimize by hashing the pool values once?
+    @inbounds for (i, ref) in enumerate(v.refs)
+        h[i] = ref == 0 ?
+               hash(NULL_MAGIC, h[i]) :
+               hash(CategoricalArrays.index(v.pool)[ref], h[i])
+    end
+    h
+end
+
+# Calculate hash for each row
+# in an efficient column-wise manner
+function hashrows(dt::AbstractDataTable)
+    res = zeros(UInt, nrow(dt))
+    for col in columns(dt)
+        hashrows_col!(res, col)
+    end
+    return res
+end
+
+# Helper function for RowGroupDict.
+# Returns a tuple:
+# 1) the number of row groups in a data frame
+# 2) vector of row hashes
+# 3) slot array for a hash map, non-zero values are
+#    the indices of the first row in a group
+# Optional group vector is set to the group indices of each row
+function row_group_slots(dt::AbstractDataTable,
+                          groups::Union{Vector{Int}, Void} = nothing)
+    @assert groups === nothing || length(groups) == nrow(dt)
+    rhashes = hashrows(dt)
+    sz = Base._tablesz(length(rhashes))
+    @assert sz >= length(rhashes)
+    szm1 = sz-1
+    gslots = zeros(Int, sz)
+    ngroups = 0
+    @inbounds for i in eachindex(rhashes)
+        # find the slot and group index for a row
+        slotix = rhashes[i] & szm1 + 1
+        gix = 0
+        probe = 0
+        while true
+            g_row = gslots[slotix]
+            if g_row == 0 # unoccupied slot, current row starts a new group
+                gslots[slotix] = i
+                gix = ngroups += 1
+                break
+            elseif rhashes[i] == rhashes[g_row] # occupied slot, check if miss or hit
+                eq = true
+                for col in columns(dt)
+                    if !isequal_colel(col, i, g_row)
+                        eq = false # miss
+                        break
+                    end
+                end
+                if eq # hit
+                    gix = groups !== nothing ? groups[g_row] : 0
+                    break
+                end
+            end
+            slotix = slotix & szm1 + 1 # check the next slot
+            probe += 1
+            probe < sz || error("Cannot find free row slot")
```

Comment:
In principle, it should never happen as we should always find a free slot in the hash after doing `sz` tries as the number of slots is no less then the number of rows. Maybe it's better to convert this line into an `@assert`.